[wip] Apply single record a time when batch fails

katyanna opened the pull request.

linki commented on controller/controller.go:
```diff
@@ -220,14 +220,26 @@ func (c *Controller) RunOnce(ctx context.Context) error {
+
+			for _, change := range changes {
+				cg := &plan.Changes{
+					Create: []*endpoint.Endpoint{change},
```
We cannot merge the deletions and updates into the creations. They are all of the same type but the respective slice defines what should happen with them, e.g. here we would now try to _create_ records that should be _deleted_.

Commit:
add single change policy

katyanna committed:
Apply SingleChangePolicy

katyanna committed:
Divide batch in half and apply them

Given an error, apply policy that returns half of an action's records
and then runs the other half if the first one fails. However might
never go through Update and Delete actions if the failing record is in
the Create slice, due to the needed actions ordering.

katyanna committed:
Enforces trial of each action

Given an error of the full batch, tries to run each action separately
and given errors, runs the half batch policies.

linki commented on controller/controller.go:
```diff
@@ -220,14 +220,33 @@ func (c *Controller) RunOnce(ctx context.Context) error {
+			deleteErr := c.Registry.ApplyChanges(ctx, deleteChanges)
+			if deleteErr != nil {
+				applyHalfActionChanges(c, ctx, deleteChanges)
+			}
```
Looks good 👍 

Maybe we can also add a textual description of the approach taken as a comment or in the PR. Maybe it helps us and other reviewers to reason about whether this splitting is safe and doesn't change the final result (apart from ignoring the failing records).

katyanna replied:
yes, yes. the current comments are more like self notes to assimilate what I'm doing in the draft, once it's ready for opening a PR upstream there will more clear documentation.

linki commented on controller/controller.go:
```diff
@@ -220,14 +220,33 @@ func (c *Controller) RunOnce(ctx context.Context) error {
+			createErr := c.Registry.ApplyChanges(ctx, createChanges)
+			// if the single action batch fails, try batch in halves
+			if createErr != nil {
+				applyHalfActionChanges(c, ctx, createChanges)
```
Since `applyHalfActionChanges` returns an error we should get it here and at least log it.

katyanna committed:
Turn update and delete only policies into simple split function
